src/textual/widgets/_data_table.py

```diff
@@ -322,82 +427,233 @@ def __init__(
+        """The number of columns to fix (prevented from scrolling)."""
         self.zebra_stripes = zebra_stripes
-        self.header_height = header_height
+        """Apply zebra effect on row backgrounds (light, dark, light, dark, ...)."""
         self.show_cursor = show_cursor
+        """Show/hide both the keyboard and hover cursor."""
         self._show_hover_cursor = False
+        """Used to hide the mouse hover cursor when the user uses the keyboard."""
+        self._update_count = 0
+        """Number of update (INCLUDING SORT) operations so far. Used for cache invalidation."""
+        self._header_row_key = RowKey()
+        """The header is a special row - not part of the data. Retrieve via this key."""
 
     @property
     def hover_row(self) -> int:
-        return self.hover_cell.row
+        """The index of the row that the mouse cursor is currently hovering above."""
+        return self.hover_coordinate.row
 
     @property
     def hover_column(self) -> int:
-        return self.hover_cell.column
+        """The index of the column that the mouse cursor is currently hovering above."""
+        return self.hover_coordinate.column
 
     @property
     def cursor_row(self) -> int:
-        return self.cursor_cell.row
+        """The index of the row that the DataTable cursor is currently on."""
+        return self.cursor_coordinate.row
 
     @property
     def cursor_column(self) -> int:
-        return self.cursor_cell.column
+        """The index of the column that the DataTable cursor is currently on."""
+        return self.cursor_coordinate.column
+
+    @property
+    def row_count(self) -> int:
+        """The number of rows currently present in the DataTable."""
+        return len(self.rows)
+
+    @property
+    def _y_offsets(self) -> list[tuple[RowKey, int]]:
+        """Contains a 2-tuple for each line (not row!) of the DataTable. Given a
+        y-coordinate, we can index into this list to find which row that y-coordinate
+        lands on, and the y-offset *within* that row. The length of the returned list
+        is therefore the total height of all rows within the DataTable."""
+        y_offsets = []
+        if self._update_count in self._offset_cache:
+            y_offsets = self._offset_cache[self._update_count]
+        else:
+            for row in self.ordered_rows:
+                y_offsets += [(row.key, y) for y in range(row.height)]
+            self._offset_cache = y_offsets
+        return y_offsets
+
+    @property
+    def _total_row_height(self) -> int:
+        """The total height of all rows within the DataTable"""
+        return len(self._y_offsets)
+
+    def update_cell(
+        self,
+        row_key: RowKey | str,
+        column_key: ColumnKey | str,
+        value: CellType,
+        *,
+        update_width: bool = False,
+    ) -> None:
+        """Update the content inside the cell with the specified row key
+        and column key.
```

Comment:
May want reformatting to confirm with [our docstring guidelines](https://github.com/Textualize/textual/wiki/Documentation-standards#class).